README.md
# :wrench: OTOS Utils 
![tag](https://img.shields.io/github/v/tag/SebastianOberschwendtner/otos-utils?color=green)
![release](https://img.shields.io/github/v/release/SebastianOberschwendtner/otos-utils?color=green)
![issues](https://img.shields.io/github/issues-raw/SebastianOberschwendtner/otos-utils)
![bugs](https://img.shields.io/github/issues/SebastianOberschwendtner/otos-utils/bug?color=red)

Utility functions for developing with the OTOS operating system.


## Software Framework
![GitHub Test Status](https://img.shields.io/github/workflow/status/SebastianOberschwendtner/otos-utils/Unit%20Test?label=test)
![GitHub top language](https://img.shields.io/github/languages/top/SebastianOberschwendtner/otos-utils?color=brightgreen)
![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/SebastianOberschwendtner/otos-utils)
![GitHub](https://img.shields.io/github/license/SebastianOberschwendtner/otos-utils)

Available Packages:
- [Font Generator](#font-generator)

The packages are briefly described in the following:

### Font Generator
The font generator tool can be used to convert *TrueType* fonts to bitmaps which can be used by the Graphics Library of *OTOS*.
Here is the the output of `run_font_converter.py -h`:
```bash
usage: run_font_generator.py [-h] --font FONT --size SIZE [SIZE ...] --output
                             OUTPUT [--version]

Generate font files for the OTOS Graphics library.

optional arguments:
  -h, --help            show this help message and exit
  --font FONT, -f FONT  Path or Name of the font file.
  --size SIZE [SIZE ...], -s SIZE [SIZE ...]
                        Size(s) of the font in pixels.
  --output OUTPUT, -o OUTPUT
                        Path to the output file.
  --version, -v         show program's version number and exit
```

CHANGELOG.md
# Changelog for OTOS - Utils
## [v1.0.0](https://github.com/SebastianOberschwendtner/OTOS-Utils/releases/tag/v1.0.0) *(2022-10-09)*

>Released by `SO`

### Release Notes:
- Initial release of *OTOS Utils*, includes:
    - `FontGenerator`: Generates font files for the *OTOS* Graphics library.


### Fixed Issues:

- n/a


test/test_FontGenerator/test_Exporter.py
# OTOS - Open Tec Operating System
# Copyright (c) 2022 Sebastian Oberschwendtner, sebastian.oberschwendtner@gmail.com
#
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
### Details
- *File:*     test_Exporter.py
- *Details:*  Python 3.9
- *Date:*     2022-10-08
- *Version:*  v1.0.0

### Author
Sebastian Oberschwendtner, :email: sebastian.oberschwendtner@gmail.com
"""
# === Modules ===
import pytest, pathlib

# === UUT ===
from utils.FontGenerator import Exporter as UUT

# === Test list ===
# ▢ File formatting:
#   ▢ Size in pixels sets the namespace correctly
# ▢ File export:
#   ▢ Copyright header is written correctly
#   ▢ Array line is written correctly
#   ▢ lookup table is written correctly
#   ▢ Font name is written correctly
#   ▢ Font data struct is written correctly


# === Fixtures ===

# === Tests ===

class Test_File_Formatting():
    """Test group to test file formatting."""
    def test_size_in_pixels_sets_namespace_correctly(self):
        """Test if the size in pixels sets the namespace correctly."""
        # Act
        name = UUT.get_namespace_for_size(8)
        # Assert
        assert len(name) == 4
        assert name == "_8px"


class Test_File_Export():
    """Test group to test file export."""
    def test_copyright_header_is_written_correctly(self, tmp_path: pathlib.Path):
        """Test if the copyright header is written correctly."""
        # Arrange
        expected = UUT._OTOS_Copyright
        file = tmp_path / "test.txt"

        # Act
        UUT.write_copyright_header(file)

        # Assert
        assert file.read_text().startswith(expected)

    def test_array_line_is_written_correctly(self, tmp_path: pathlib.Path):
        """Test if the lookup table is written correctly."""
        # Arrange
        expected = "0x00, 0x01, 0x02, 0x03, // 0x00\n"

        # Act
        line = UUT.get_array_line(0, [0,1,2,3])

        # Assert
        assert line.startswith(expected)

    def test_lookup_table_preamble(self, tmp_path: pathlib.Path):
        """Test if the lookup table is written correctly."""
        # Arrange
        expected = "\n"
        expected += "#ifndef TESTFONT_H_\n"
        expected += "#define TESTFONT_H_\n"
        expected += "\n"
        expected += "// === Includes ===\n"
        expected += "#include \"font_base.h\"\n"
        expected += "\n"
        expected += "namespace Font\n"
        expected += "{\n"

        # Act
        file = tmp_path / "test.txt"
        file.touch()
        UUT.write_lookup_table_preamble(file, "TestFont")

        # Assert
        assert file.read_text().startswith(expected)

    def test_lookup_table_begin(self, tmp_path: pathlib.Path):
        """Test if the lookup table is written correctly."""
        # Arrange
        expected = "    /**\n"
        expected += "     * @brief Ascii font lookup table\n"
        expected += "     * @details width: 12 px, height: 20 px\n"
        expected += "     */\n"
        expected += "    constexpr unsigned char Lookup_TestFont_20px[] = {\n"

        # Act
        file = tmp_path / "test.txt"
        file.touch()
        UUT.write_lookup_table_begin(file, "TestFont", (12, 20))

        # Assert
        assert file.read_text().startswith(expected)

    def test_lookup_table_end(self, tmp_path: pathlib.Path):
        """Test if the lookup table is written correctly."""
        # Arrange
        expected = "    };\n"
        expected += "\n"
        expected += "    // === Font Information ===\n"
        expected += "    // TestFont: 20px\n"
        expected += "    namespace _20px\n"
        expected += "    {\n"
        expected += "        constexpr Font::Base_t TestFont = {\n"
        expected += "            .data = Lookup_TestFont_20px,\n"
        expected += "            .width_px = 12,\n"
        expected += "            .height_px = 20,\n"
        expected += "            .stride = 3};\n"
        expected += "    };\n"

        # Act
        file = tmp_path / "test.txt"
        file.touch()
        UUT.write_lookup_table_end(file, "TestFont", (12, 20), 3)

        # Assert
        assert file.read_text().startswith(expected)

    def test_finalizing_file(self, tmp_path: pathlib.Path):
        """Test if the lookup table is written correctly."""
        # Arrange
        expected = "};\n"
        expected += "#endif /* TESTFONT_H_ */"

        # Act
        file = tmp_path / "test.txt"
        file.touch()
        UUT.finalize_file(file, "TestFont")

        # Assert
        assert file.read_text().startswith(expected)

test/test_FontGenerator/test_BitConverter.py
# OTOS - Open Tec Operating System
# Copyright (c) 2022 Sebastian Oberschwendtner, sebastian.oberschwendtner@gmail.com
#
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
### Details
- *File:*     test_BitConverter.py
- *Details:*  Python 3.9
- *Date:*     2022-10-08
- *Version:*  v1.0.0

### Author
Sebastian Oberschwendtner, :email: sebastian.oberschwendtner@gmail.com
"""
# === Modules ===
import pytest, pathlib

# === UUT ===
from utils.FontGenerator import BitConverter as UUT

# === Test list ===
# ▢ Create a canvas according to the desired font size and width
# ▢ Draw the character on the canvas
# ▢ Preview the character
# ▢ Convert the canvas to a numpy array

# === Fixtures ===
@pytest.fixture
def Path_Test_Font() -> pathlib.Path:
    yield pathlib.Path("./test/Stubs/DelugiaMonoPL.ttf")

# === Tests ===

class Test_Basic_Canvas():
    """Test group to test the basic canvas creation."""
    def test_create_canvas(self):
        """Test if a canvas can be created."""
        # Arrange
        # Act
        canvas = UUT.create_canvas(8, 8)
        # Assert
        assert canvas.size == (8, 8)
        assert canvas.mode == "1"

    def test_get_max_width(self, Path_Test_Font: pathlib.Path):
        """Test if the max width is returned correctly."""
        # Arrange
        font = UUT.ImageFont.truetype(str(Path_Test_Font), 32)
        # Act
        max_width = UUT.get_max_width(font)
        # Assert
        assert max_width == 20

    def test_get_max_offset(self, Path_Test_Font: pathlib.Path):
        """Test if the max width is returned correctly."""
        # Arrange
        font = UUT.ImageFont.truetype(str(Path_Test_Font), 32)
        # Act
        max_offset = UUT.get_max_offset(font)
        # Assert
        assert max_offset == 6

    def test_converting_pixel_sequence(self):
        """Test if the pixel sequence is converted correctly."""
        # Arrange
        pixel_sequence = [0, 0, 0, 0, 0, 0, 0, 0]
        # Act
        converted = UUT.convert_pixel_sequence(pixel_sequence)
        # Assert
        assert converted == 0

        # Arrange
        pixel_sequence = [1, 0, 0, 0, 0, 0, 0, 0]
        # Act
        converted = UUT.convert_pixel_sequence(pixel_sequence)
        # Assert
        assert converted == 1

        # Arrange
        pixel_sequence = 8*[1,]
        # Act
        converted = UUT.convert_pixel_sequence(pixel_sequence)
        # Assert
        assert converted == 255

class Test_Font_Converter():
    """Test group to test the font converter."""
    def test_convert_character_8px(self, Path_Test_Font: pathlib.Path):
        """Test if the char is converted correctly."""
        # Arrange
        Converter = UUT.FontConverter(str(Path_Test_Font), 8)

        # Act
        char = Converter.convert_character(ord("A"))

        # Assert
        assert len(char) == 5
        assert char[0] == 0b00100000
        assert char[1] == 0b00011100
        assert char[2] == 0b00010010
        assert char[3] == 0b00111100
        assert char[4] == 0b00100000

    def test_convert_character_16px(self, Path_Test_Font: pathlib.Path):
        """Test if the char is converted correctly."""
        # Arrange
        Converter = UUT.FontConverter(str(Path_Test_Font), 16)

        # Act
        char = Converter.convert_character(ord("A"))

        # Assert
        assert len(char) == 20
        assert char[0] == 0b00010000
        assert char[1] == 0b00000000
        assert char[2] == 0b00011110
        assert char[3] == 0b00000000
        assert char[4] == 0b00000111
        assert char[5] == 0b11100000
        assert char[6] == 0b00000011
        assert char[7] == 0b01111110
        assert char[8] == 0b00000011
        assert char[9] == 0b00000110

    def test_convert_character_24px(self, Path_Test_Font: pathlib.Path):
        """Test if the char is converted correctly."""
        # Arrange
        Converter = UUT.FontConverter(str(Path_Test_Font), 24)

        # Act
        char = Converter.convert_character(ord("A"))

        # Assert
        assert len(char) == 45
        assert char[0] == 0b00000000
        assert char[1] == 0b00000000
        assert char[2] == 0b00000000
        assert char[3] == 0b00000111
        assert char[4] == 0b00000000
        assert char[5] == 0b00000000
        assert char[6] == 0b00000111
        assert char[7] == 0b11110000
        assert char[8] == 0b00000000

    def test_get_fontname(self, Path_Test_Font: pathlib.Path):
        """Test if the font name is returned correctly."""
        # Arrange
        Converter = UUT.FontConverter(str(Path_Test_Font), 8)

        # Act
        fontname = Converter.fontname

        # Assert
        assert fontname == "DelugiaPLMono"

    def test_open_invalid_font(self):
        """Test correct handling of invalid font files."""
        # Arrange
        # Act
        with pytest.raises(FileNotFoundError):
            Converter = UUT.FontConverter("invalid_font.ttf", 8)

test/test_FontGenerator/test_FontGenerator.py
# OTOS - Open Tec Operating System
# Copyright (c) 2022 Sebastian Oberschwendtner, sebastian.oberschwendtner@gmail.com
#
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
### Details
- *File:*     test_FontGenerator.py
- *Details:*  Python 3.9
- *Date:*     2022-10-08
- *Version:*  v1.0.0

### Author
Sebastian Oberschwendtner, :email: sebastian.oberschwendtner@gmail.com
"""
# === Modules ===
import pytest, pathlib

# === UUT ===
from utils import FontGenerator as UUT

# === Test list ===
# ▢ Font Data Type:
#   ▢ has name property
#   ▢ has size property
#   ▢ has width property
#   ▢ has stride property
#   ▢ has data property with 256 entries
# ▢ Font Class:
#   ▢ has a data container
#   ▢ has a converter

# === Fixtures ===
@pytest.fixture
def FontConverter_Mock(mocker):
    """Mock the FontConverter class."""
    _mock = mocker.patch.multiple(
        "utils.FontGenerator.BitConverter.FontConverter",
        __init__=mocker.DEFAULT,
        convert_character=mocker.DEFAULT,
        fontname = "TestFont",
        height_px = 8,
        width_px = 5,
    )
    _mock["__init__"].return_value = None
    _mock["convert_character"].return_value = [0, 1, 2, 3, 4]
    yield _mock

# === Tests ===


class Test_Font_Data_Type():
    """Test group to test the font data type."""
    def test_has_name_property(self):
        """Test if the font data type has a name property."""
        # Arrange
        font = UUT.FontData()
        # Assert attribute
        assert hasattr(font, "name")
        # Act
        font.name = "Test"
        # Assert
        assert font.name == "Test"

    def test_has_size_property(self):
        """Test if the font data type has a size property."""
        # Arrange
        font = UUT.FontData()
        # Assert attribute
        assert hasattr(font, "size")
        # Act
        font.size = 8
        # Assert
        assert font.size == 8

    def test_has_width_property(self):
        """Test if the font data type has a width property."""
        # Arrange
        font = UUT.FontData()
        # Assert attribute
        assert hasattr(font, "width")
        # Act
        font.width = 8
        # Assert
        assert font.width == 8

    def test_has_stride_property(self):
        """Test if the font data type has a stride property."""
        # Arrange
        font = UUT.FontData()
        # Assert attribute
        assert hasattr(font, "stride")
        # Act
        font.stride = 8
        # Assert
        assert font.stride == 8

    def test_has_data_property_with_256_entries(self):
        """Test if the font data type has a data property with 256 entries."""
        # Arrange
        font = UUT.FontData()
        # Assert attribute
        assert hasattr(font, "data")
        # Assert
        assert len(font.data) == 256

class Test_Font_Class():
    """Test group to test the font class."""
    def test_init(self, FontConverter_Mock):
        """Test if the font class has a data container."""
        # Arrange
        font_source = pathlib.Path("test/test_FontGenerator/test_font.ttf")
        font = UUT.Font(font_source, 12)
        # Assert attribute
        assert hasattr(font, "data")
        assert hasattr(font, "converter")
        # Assert
        assert isinstance(font.data, UUT.FontData)
        FontConverter_Mock["__init__"].assert_called_once_with(str(font_source), 12)

    def test_convert(self, FontConverter_Mock):
        """Test the converting of a font."""
        # Arrange
        font_source = pathlib.Path("test/test_FontGenerator/test_font.ttf")
        font = UUT.Font(font_source, 12)
        # Act
        font.convert()
        # Assert
        assert font.data.name == "TestFont"
        assert font.data.size == 8 #FontConverter_Mock["height_px"]
        assert font.data.width == 5 #FontConverter_Mock["width_px"]
        assert font.data.stride == 1
        assert FontConverter_Mock["convert_character"].call_count == 256
        assert font.data.data[0] == [0,1,2,3,4]

utils/FontGenerator/Exporter.py
# OTOS - Open Tec Operating System
# Copyright (c) 2022 Sebastian Oberschwendtner, sebastian.oberschwendtner@gmail.com
#
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
### Details
- *File:*     FontGenerator/Exporter.py
- *Details:*  Python 3.9
- *Date:*     2022-10-08
- *Version:*  v1.0.0

## Description
Exporter for the font generator.

### Author
Sebastian Oberschwendtner, :email: sebastian.oberschwendtner@gmail.com
"""
# === Modules ===
from . import pathlib

# === Functions ===


def get_namespace_for_size(size: int) -> str:
    """Converts the font size to a namespace.

    Args:
        size (int): 1x1 [px] The font size in pixels.

    Returns:
        str: 1xn [str] The namespace for the font size.

    ---
    """
    return f"_{size}px"


def write_copyright_header(file: pathlib.Path):
    """Writes the copyright header to the file.

    Args:
        file (pathlib.Path): 1x1 [-] The file to write to.

    ---
    """
    # Write the header
    file.write_text(_OTOS_Copyright)


def get_array_line(line_number: int, array: list) -> str:
    """Get the line of an array as a string.

    Args:
        line_number (int): 1x1 [-] The line number of the array in the file.
        array (list): 1xn [-] The array to write.

    Returns:
        str: 1x1 [-] The line of the array.

    ---
    """
    # Convert every element in the array to a hex string
    line_string = ""
    for Item in array:
        line_string += f"{Item:#04x}, "

    # End the line with the line number as comment
    if chr(line_number).isprintable() and line_number != 0x5c:
        line_string += f"// {line_number:#04x}: {chr(line_number)}" + "\n"
    else:
        line_string += f"// {line_number:#04x}" + "\n"

    # Return the line
    return line_string


def write_lookup_table_preamble(file: pathlib.Path, font_name: str):
    """Write the preamble of the lookup table.

    Args:
        file (pathlib.Path): 1x1 [-] The file to write to.
        font_name (str): 1x1 [-] The name of the font.

    ---
    """
    # Write the beginning of the lookup table
    with open(file, "a") as File:
        File.write(_OTOS_LookUp_Preamble.format(Name_Upper=font_name.upper()))
        File.write(r"{" + "\n")


def write_lookup_table_begin(file: pathlib.Path, font_name: str, size: tuple):
    """Write the beginning of the lookup table.

    Args:
        file (pathlib.Path): 1x1 [-] The file to write to.
        font_name (str): 1x1 [-] The name of the font.
        size (tuple): 1x2 [px] The (width, height) of the font.

    ---
    """
    # Write the beginning of the lookup table
    with open(file, "a") as File:
        File.write(
            _OTOS_LookUp_Begin.format(Name=font_name, Width=size[0], Height=size[1])
        )
        File.write(r"{" + "\n")


def write_lookup_table_end(
    file: pathlib.Path, font_name: str, size: tuple, stride: int
):
    """Write the end of the lookup table.

    Args:
        file (pathlib.Path): 1x1 [-] The file to write to.
        font_name (str): 1x1 [-] The name of the font.
        size (tuple): 1x2 [px] The (width, height) of the font.
        stride (int): 1x1 [-] The stride of the font.

    ---
    """
    # Write the end of the lookup table
    with open(file, "a") as File:
        File.write(
            _OTOS_LookUp_End.format(
                Namespace=get_namespace_for_size(size[1]),
                Name=font_name,
                Width=size[0],
                Height=size[1],
                Stride=stride,
            )
        )

def finalize_file(file: pathlib.Path, font_name: str):
    """Finalize the file.

    Args:
        file (pathlib.Path): 1x1 [-] The file to write to.
        font_name (str): 1x1 [-] The name of the font.

    ---
    """
    # Write the end of the file
    with open(file, "a") as File:
        File.write(f"}};\n#endif /* {font_name.upper()}_H_ */")


# === Constants ===
_OTOS_Copyright: str = """/**
 * OTOS - Open Tec Operating System
 * Copyright (c) 2022 Sebastian Oberschwendtner, sebastian.oberschwendtner@gmail.com
 *
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
 """

_OTOS_LookUp_Preamble: str = """
#ifndef {Name_Upper}_H_
#define {Name_Upper}_H_

// === Includes ===
#include "font_base.h"

namespace Font
"""

_OTOS_LookUp_Begin: str = """    /**
     * @brief Ascii font lookup table
     * @details width: {Width:d} px, height: {Height:d} px
     */
    constexpr unsigned char Lookup_{Name}_{Height:d}px[] = """

_OTOS_LookUp_End: str = """    }};

    // === Font Information ===
    // {Name}: {Height:d}px
    namespace {Namespace}
    {{
        constexpr Font::Base_t {Name} = {{
            .data = Lookup_{Name}_{Height:d}px,
            .width_px = {Width:d},
            .height_px = {Height:d},
            .stride = {Stride:d}}};
    }};
"""


utils/FontGenerator/BitConverter.py
# OTOS - Open Tec Operating System
# Copyright (c) 2022 Sebastian Oberschwendtner, sebastian.oberschwendtner@gmail.com
#
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
### Details
- *File:*     FontGenerator/BitConverter.py
- *Details:*  Python 3.9
- *Date:*     2022-10-08
- *Version:*  v1.0.0

## Description
Converts a TTF font to a bitmap which can be used by OTOS.

### Author
Sebastian Oberschwendtner, :email: sebastian.oberschwendtner@gmail.com
"""
# === Modules ===
from PIL import Image, ImageFont, ImageDraw

# === Functions ===

def create_canvas(height_px: int, width_px: int) -> Image:
    """Creates a canvas for the font.
    Args:
        height_px (int): 1x1 [px] The height of the canvas.
        width_px (int): 1x1 [px] The width of the canvas.
    Returns:
        PIL.Image: 1x1 [-] The canvas.


    ---
    """
    return Image.new("1", (width_px, height_px), 0)

def get_max_width(font: ImageFont.FreeTypeFont) -> int:
    """Get the maximum required width of the font in pixels.

    Args:
        font (ImageFont.FreeTypeFont): 1x1 [-] The current font type with requested font size.

    Returns:
        int: 1x1 [px] The maximum width of the font in pixels.

    ---
    """
    # Get the maximum width of the font
    max_width = 0
    for iChar in range(256):
        left, top, right, bottom = font.getbbox(chr(iChar))
        width = right - left
        if width > max_width:
            max_width = width

    return max_width

def get_max_offset(font: ImageFont.FreeTypeFont) -> int:
    """Get the maximum offset in y direction to fit all characters.

    Args:
        font (ImageFont.FreeTypeFont): 1x1 [-] The current font type with requested font size.

    Returns:
        int: 1x1 [px] The maximum Y offset of the font in pixels.

    ---
    """
    # Get the maximum width of the font
    max_offset = 0
    for iChar in range(256):
        left, top, right, bottom = font.getbbox(chr(iChar))
        if bottom > max_offset:
            max_offset = top

    return max_offset

def convert_pixel_sequence(sequence) -> int:
    """Converts a pixel sequence to a byte.

    Args:
        sequence (list): 1x8 [px] The pixel sequence.

    Returns:
        int: 1x1 [-] The converted byte.

    ---
    """
    # Convert the sequence to a byte
    byte = 0
    for i in range(8):
        byte |= sequence[i] << i

    return byte

# === Classes ===


class FontConverter:
    """Converts a TTF font to a bitmap which can be used by OTOS.

    ---
    """
    # === Properties ===
    @property
    def fontname(self) -> str:
        """Gets the name of the font.

        Returns:
            str: 1x1 [-] The name of the font.

        ---
        """
        _name = self.font.getname()[0]
        return _name.replace(" ", "")

    @property
    def height_px(self) -> int:
        return self._height_px

    @property
    def width_px(self) -> int:
        return self._width_px

    # === Constructor ===
    def __init__(self, font_path: str, font_size: int):
        """Creates a new font converter.

        Args:
            font_path (str): 1x1 [-] The path to the font file.
            font_size (int): 1x1 [px] The font size in pixels.

        ---
        """
        # Save the font path
        self.font_path = font_path
        self._height_px = font_size

        # Load the font
        try:
            self.font = ImageFont.truetype(font_path, font_size)
        except:
            raise FileNotFoundError("Font file not found.")

        # Get the maximum width of the font
        self._width_px = get_max_width(self.font)

    # === Methods ===
    def convert_character(self, character: int) -> list:
        """Converts a character to a bitmap.

        Args:
            character (int): 1x1 [-] The character to convert. 

        Returns:
            list: 1x1 [-] The bitmap.

        ---
        """
        # Create the canvas
        y_offset = get_max_offset(self.font)
        canvas = create_canvas(self.height_px, self.width_px)

        # Draw the font
        draw = ImageDraw.Draw(canvas)
        draw.text((0, -y_offset), chr(character), font=self.font, fill=1)

        # Convert the canvas to a bitmap
        bytes_y = self.height_px // 8
        bitmap = []
        for x in range(self.width_px):
            for y in range(bytes_y):
                # Get the pixel sequence
                sequence = []
                for i in range(8):
                    sequence.append(canvas.getpixel((x, (bytes_y-1-y)*8+i)))

                # Convert the sequence to a byte
                byte = convert_pixel_sequence(sequence)

                # Append the byte to the bitmap
                bitmap.append(byte)

        return bitmap

utils/FontGenerator/__init__.py
#!python
# OTOS - Open Tec Operating System
# Copyright (c) 2022 Sebastian Oberschwendtner, sebastian.oberschwendtner@gmail.com
#
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
### Details
- *File:*     FontGenerator/__init__.py
- *Details:*  Python 3.9
- *Date:*     2022-10-08
- *Version:*  v1.0.0

## Description
This utility module can be used to generate font files for the *OTOS*
Graphics library.

### Author
Sebastian Oberschwendtner, :email: sebastian.oberschwendtner@gmail.com
"""
# === Modules ===
import pathlib, math
from . import BitConverter, Exporter

# === Package Information ===
__author__ = "Sebastian Oberschwendtner"
__version__ = "1.0.0"
__all__ = ["BitConverter", "Exporter"]

# === Functions ===

# === Classes ===
class FontData:
    """Data type to store font data."""

    def __init__(self):
        """Constructor of the font data type."""
        self.name = ""
        self.size = 0
        self.width = 0
        self.stride = 0
        self.data = [
            [],
        ] * 256


class Font:
    """Class to generate font files."""

    def __init__(self, font_file: pathlib.Path, font_size: int):
        """Constructor of the font class."""
        self.data: FontData = FontData()
        self.converter: BitConverter.FontConverter = BitConverter.FontConverter(
            str(font_file), font_size
        )

    def convert(self):
        """Convert the font file to a font file."""
        # Assign meta data
        self.data.name = self.converter.fontname
        self.data.size = self.converter.height_px
        self.data.width = self.converter.width_px
        self.data.stride = int(math.ceil(self.converter.height_px / 8))

        # Convert the characters
        for i in range(256):
            self.data.data[i] = self.converter.convert_character(i)


class Fonts:
    """Class to generate multiple font files."""

    def __init__(self, font_file: pathlib.Path, font_sizes: list):
        """Constructor of the fonts class."""
        self.fonts = []
        for iSize in font_sizes:
            self.fonts.append(Font(font_file, iSize))

    def convert(self):
        """Convert all fonts."""
        for font in self.fonts:
            font.convert()

    def export(self, export_path: pathlib.Path):
        """Export the font file."""
        # Create file
        _name = self.fonts[0].data.name
        file = export_path / f"{_name}.h"

        # Write Copyright
        Exporter.write_copyright_header(file)

        # Write autogenerated warning
        with open(file, "a") as f:
            f.write(
                f"\n/**\n * @attention\n * This file is autogenerated by {__name__} - {__version__}.\n * You should probably not edit this file manually.\n */\n"
            )

        # Write lookup preamble
        Exporter.write_lookup_table_preamble(file, _name)

        for iFont in self.fonts:
            # Write lookup table begin
            Exporter.write_lookup_table_begin(
                file, _name, (iFont.data.width, iFont.data.size)
            )

            # Write lookup table
            with open(file, "a") as f:
                for iChar in range(256):
                    line = Exporter.get_array_line(iChar, iFont.data.data[iChar])
                    f.write(8 * " " + line)

            # Write lookup table end
            Exporter.write_lookup_table_end(
                file, _name, (iFont.data.width, iFont.data.size), iFont.data.stride
            )

        # Finalize file
        Exporter.finalize_file(file, _name)
